File: stack/minstack.py
class MinStack:

    def __init__(self):
        self.stack = []
        
    def push(self, val: int) -> None:
        min_val = min(val, self.getMin())
        '''
        stack has val and min_val
        val is the value which need to be pushed into thestack
        min_val is the value which is minimum till that point
        irrespective of push value
        '''
        self.stack.append((val, min_val))
        
    def pop(self) -> None:
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1][0]

    def getMin(self) -> int:
        if self.stack:
            return self.stack[-1][1]
        return float('inf')


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: stack/minstack.java
package stack;

import java.util.Stack;

public class minstack {
    private Stack<int[]> stack;

    public MinStack() {
        stack = new Stack<>();
    }
    
    public void push(int val) {
        int minVal = Math.min(val, this.getMin());
        stack.push(new int[]{val, minVal});
    }
    
    public void pop() {
        stack.pop();
    }
    
    public int top() {
        return stack.peek()[0];
    }
    
    public int getMin() {
        if(!stack.empty()){
            return stack.peek()[1];
        }
        return Integer.MAX_VALUE;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */